examples/gnrc_tcp_cli/main.c

```diff
@@ -0,0 +1,231 @@
+/*
+ * Copyright (C) 2015 Simon Brummer
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+#include <stdio.h>
+#include <errno.h>
+#include "net/af.h"
+#include "random.h"
+#include "net/gnrc/ipv6.h"
+#include "net/gnrc/tcp.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+#ifndef TARGET_ADDR
+#define TARGET_ADDR "fe80::ccea:ebff:fe12:f132"
+#endif
+
+#ifndef TARGET_PORT
+#define TARGET_PORT 80 /* Target Port */
+#endif
+
+#ifndef CONNS
+#define CONNS 1        /* Number of parallel connections */
+#endif
+
+#ifndef NBYTE          /* Number of parallel connections */
+#define NBYTE (2048)   /* Amount of data to transmit */
+#endif
+
+#ifndef CYCLES
+#define CYCLES 10      /* Number of repeated Cycles per thread */
+#endif
+
+uint8_t bufs[CONNS][NBYTE];
+uint8_t stacks[CONNS][THREAD_STACKSIZE_DEFAULT + THREAD_EXTRA_STACKSIZE_PRINTF];
+
+void *cli_thread(__attribute__((unused)) void *arg);
+
+int main(void)
+{
+    kernel_pid_t ifs[GNRC_NETIF_NUMOF];
+
+    /* Disable 15.4 Retransmissions on the first network device */
+    netopt_enable_t opt = NETOPT_DISABLE;
+    gnrc_netif_get(ifs);
+    gnrc_netapi_set(ifs[0], NETOPT_AUTOACK, 0, &opt, sizeof(opt));
+    gnrc_netapi_set(ifs[0], NETOPT_ACK_REQ, 0, &opt, sizeof(opt));
+
+    printf("\nStarting Client Threads. Config is: TARGET_ADDR=%s, TARGET_PORT=%d, ", TARGET_ADDR, TARGET_PORT);
+    printf("CONNS=%d, NBYTE=%d, CYCLES=%d\n\n", CONNS, NBYTE, CYCLES );
+
+    /* Start Connection Handling Threads */
+    for (int i = 0; i < CONNS; i += 1) {
+        thread_create((char *)stacks[i], sizeof(stacks[i]), THREAD_PRIORITY_MAIN, 0, cli_thread,
+                      (void *)i, NULL);
+    }
+
+    return 0;
+}
+
+void *cli_thread(__attribute__((unused)) void *arg)
+{
+    /* Test Program variables */
+    int32_t tid = (int32_t) arg;
+    uint32_t rcvd = 0;
+    uint32_t sent = 0;
+    uint32_t cycles = 0;
+    uint32_t cycles_ok = 0;
+    uint32_t cycles_faulty = 0;
+
+    /* TCP specific variables */
+    ipv6_addr_t local_addr;
+    ipv6_addr_t peer_addr;
+    uint16_t local_port;
+    uint16_t peer_port;
+    gnrc_tcp_tcb_t tcb;
+
+    /* Target Peer Address Information */
+    ipv6_addr_t target_addr;
+    uint16_t target_port;
```

Comment:
local and peer act as address buffer (thats why these are cleared before usage). In a sock implementation these would vanish in the sock structure. 

Target specifies the actual Peer Host